libs/rand/random.go

```diff
@@ -2,161 +2,37 @@ package rand
 
 import (
 	crand "crypto/rand"
+	"encoding/binary"
 	mrand "math/rand"
-	"time"
-
-	tmsync "github.com/tendermint/tendermint/libs/sync"
 )
 
 const (
 	strChars = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz" // 62 characters
 )
 
-// Rand is a prng, that is seeded with OS randomness.
-// The OS randomness is obtained from crypto/rand, however none of the provided
-// methods are suitable for cryptographic usage.
-// They all utilize math/rand's prng internally.
+// NewRand returns a prng, that is seeded with OS randomness.
+// The OS randomness is obtained from crypto/rand, however, like with any math/rand.Rand
+// object none of the provided methods are suitable for cryptographic usage.
 //
-// All of the methods here are suitable for concurrent use.
-// This is achieved by using a mutex lock on all of the provided methods.
-type Rand struct {
-	tmsync.Mutex
-	rand *mrand.Rand
-}
-
-var grand *Rand
-
-func init() {
-	grand = NewRand()
-	grand.init()
-}
-
-func NewRand() *Rand {
-	rand := &Rand{}
-	rand.init()
-	return rand
-}
-
-func (r *Rand) init() {
-	bz := cRandBytes(8)
-	var seed uint64
-	for i := 0; i < 8; i++ {
-		seed |= uint64(bz[i])
-		seed <<= 8
-	}
-	r.reset(int64(seed))
-}
-
-func (r *Rand) reset(seed int64) {
-	r.rand = mrand.New(mrand.NewSource(seed)) // nolint:gosec // G404: Use of weak random number generator
-}
-
-//----------------------------------------
-// Global functions
-
-func Seed(seed int64) {
-	grand.Seed(seed)
+// As this returns a math/rand.Rand instance all methods on
+// that object are suitable for concurrent use.
+func NewRand() *mrand.Rand {
+	var seed int64
+	binary.Read(crand.Reader, binary.BigEndian, &seed)
+	return mrand.New(mrand.NewSource(seed))
```

Comment:
I think this actually isn't thread-safe, we instead need to use mrand.Seed for it to be thread-safe. (Per comment on https://golang.org/pkg/math/rand/#Seed, and NewSource not being thread safe comment on https://golang.org/pkg/math/rand/#NewSource)

Reply:
Good catch! It's probably best to make this function not return anything but only call math/rand.Seed instead. So wherever a freshly seeded math/rand.Rand was used, I'll just reseed the global one lives internally in math/Rand by calling this function instead. Tomorrow...

Reply:
So, I've added a comment that clarifies that this isn't thread-safe (https://github.com/tendermint/tendermint/pull/6364/commits/5bd124ff36b23ee1cedded0175c3585faacc2b24) and added a convenience function to reseed math/rand's global Source (https://github.com/tendermint/tendermint/pull/6364/commits/7845a53a0c9c0265898329ecb8f3d733ca891a83).